modules/apps/document-library/document-library-document-conversion/src/main/java/com/liferay/document/library/document/conversion/internal/DocumentConversionImpl.java

```diff
@@ -133,19 +140,82 @@ else if (!inputDocumentFormat.isExportableTo(outputDocumentFormat)) {
 		UnsyncByteArrayOutputStream unsyncByteArrayOutputStream =
 			new UnsyncByteArrayOutputStream();
 
-		DocumentConverter documentConverter = _getDocumentConverter();
+		AtomicBoolean isDone = new AtomicBoolean(false);
+		long start = System.currentTimeMillis();
 
-		documentConverter.convert(
-			inputStream, inputDocumentFormat, unsyncByteArrayOutputStream,
-			outputDocumentFormat);
+		long timeout = Math.max(
+			PropsValues.DL_FILE_ENTRY_PREVIEW_GENERATION_TIMEOUT_GHOSTSCRIPT,
+			PropsValues.DL_FILE_ENTRY_PREVIEW_GENERATION_TIMEOUT_PDFBOX);
 
-		FileUtil.write(
-			file, unsyncByteArrayOutputStream.unsafeGetByteArray(), 0,
-			unsyncByteArrayOutputStream.size());
+		_noticeableExecutorService = _portalExecutorManager.getPortalExecutor(
+			DocumentConversionImpl.class.getName());
 
-		inputStream.close();
+		InputStream finalInputStream = inputStream;
 
-		return file;
+		Future<?> future = _noticeableExecutorService.submit(
+			() -> {
+				try {
+					DocumentConverter documentConverter =
+						_getDocumentConverter();
+
+					documentConverter.convert(
+						finalInputStream, inputDocumentFormat,
+						unsyncByteArrayOutputStream, outputDocumentFormat);
+
+					FileUtil.write(
+						file, unsyncByteArrayOutputStream.unsafeGetByteArray(),
+						0, unsyncByteArrayOutputStream.size());
+
+					isDone.set(true);
+
+					if (_log.isInfoEnabled()) {
+						_log.info(
+							StringBundler.concat(
+								"Conversion from ",
+								inputDocumentFormat.getName(), " to ",
+								outputDocumentFormat.getName(), " in ",
+								System.currentTimeMillis() - start, " ms"));
+					}
+				}
+				catch (IOException ioException) {
+					throw new RuntimeException(ioException);
+				}
+				finally {
+					try {
+						finalInputStream.close();
+					}
+					catch (IOException ioException) {
+						_log.error("Unable to close input stream", ioException);
+					}
+				}
+			});
+
+		try {
+			future.get(timeout, TimeUnit.SECONDS);
+
+			futures.put(id, future);
+		}
+		catch (TimeoutException timeoutException) {
+			String errorMessage =
+				"Timeout when converting for " + file.getPath();
+
+			if (future.cancel(true)) {
+				errorMessage += " resulted in a canceled timeout for " + future;
+			}
+
+			_log.error(errorMessage);
+
+			throw timeoutException;
+		}
+		finally {
+			_noticeableExecutorService.shutdownNow();
+		}
+
+		if (isDone.get()) {
+			return file;
+		}
+
+		throw new TimeoutException("Process timed out after timeout");
```

Comment:
This is weird. If we got here, it means that the future didn't timeout (it either succeded or failed, but under the expected time frame). See my comments above about separating success/failure/timeout.